### ISSUE TITLE
rust: Allow building dependencies twice with `native: true` and `native: false`

### ISSUE THREAD

**Issue by sdroege** [mesonbuild/meson#11744]
See https://github.com/mesonbuild/meson/issues/11702

This is required as `proc-macro` target must be built with `native: true`, but a `proc-macro` target might have the same dependencies as a non-`proc-macro` target. In that situation this dependency needs to be compiled twice, which is currently not possible.

**Comment by xclaesse** [comment]
wondering if proc-macro could be simple `declare_dependency(sources: 'mymacro.rs')` that we pass to different targets so it get rebuilt as part of each target that uses it. Unless you think that would increase significantly the build time?

**Comment by sdroege** [comment]
Apart from increasing build times, it also wouldn't solve the problem. `mymacro.rs` might use the same Rust dependency as some other non-native Rust target, so you would have to also handle that dependency in such a way.

**Comment by xclaesse** [comment]
Ah they can have dependencies other than proc-macro? I thought it was more like header files.

**Comment by sdroege** [comment]
No, they're normal Rust code that is compiled into a shared library that can then be loaded by the compiler :)

**Comment by sdroege** [comment]
@xclaesse Maybe another solution here would be to simply convert cargo projects to two dependencies (or maybe even two subprojects?). One would be compiled natively, the other not. And proc-macro targets would have to depend on the native ones while other targets would depend on the others.

**Comment by xclaesse** [comment]
Being able to build a subprojects twice for build and host machine is something we wanted for a while. I think that would be a solution here but would require more work.

**Comment by sdroege** [comment]
If there's more use-cases of this than Rust proc-macros that makes sense, sure :)

**Comment by xclaesse** [comment]
Another use case is glib, we could build tools like glib-compile-ressources for build machine, that would require building the libs too.

**Comment by bonzini** [comment]
How would you handle naming conflicts? For the specific case of the Cargo interpreter I am adding a suffix to the `native: true` case, but that does not generalize well to a hypothetical `subproject(..., native: true)`.
